### ISSUE TITLE
[2.5.9/master] SIGSEGV on move to unknown mailbox

### ISSUE THREAD

**Issue by MASHtm** [cyrusimap/cyrus-imapd#34]
I updated to 2.5.9 on my murder frontends today and got several coredumps. Mostly caused by roundcube. After some debugging I recognized that this happens if roundcube has "INBOX.Trash" configured as trash but the user deleted the folder somehow and it is not existing anymore while still pushing the "delete" button.

The difference between 2.4 and 2.5/master seems this part:
```
--- imapd.c.orig	2016-09-29 01:16:26.282596884 +0200
+++ imapd.c	2016-09-29 01:17:26.993901585 +0200
@@ -5290,7 +5290,7 @@
     }
 
     /* need permission to delete from source if it's a move */
-    if (ismove && !(imapd_index->myrights & ACL_EXPUNGE))
+    if (!r && ismove && !(imapd_index->myrights & ACL_EXPUNGE))
 	r = IMAP_PERMISSION_DENIED;
 
     /* local mailbox -> local mailbox */
```

The backtraces all showed that imapd_index is NULL causing the segfault. Since everything else in cmd_copy is "!r" protected in case the mailbox is not found I think it's needed for the permission check as well.

I'm rebuilding with this change now ...

**Comment by brong** [comment]
That shouldn't make any difference.  It's the right thing to do, but:

    /* need permission to delete from source if it's a move */
    if (ismove && !(imapd_index->myrights & ACL_EXPUNGE))
        r = IMAP_PERMISSION_DENIED;

All it's doing is setting 'r'.  Either way, 'r' should still be set after this point if there was an earlier error.

Mind you, that entire codebase should be doing:

if (r) goto done;

rather than the if (!r && ...)

Anyway, let's think about imapd_index.  This is set when a mailbox is selected.  There's no check that it's set, but COPY should only work when you have a selected mailbox.

So the bug is probably way up in cmdloop:

            else if (!strcmp(cmd.s, "Copy")) {
                if (!imapd_index && !backend_current) goto nomailbox;

So we have a current backend, but no current imapd_index I guess.

    if (!r && backend_current) {

And then we get 'r' from mlookup, so it doesn't do the proxy case, and we're screwed :(

This is a flat out code bug.  There's a path that gets us in here thinking it's a local copy even though the source mailbox is actually remote.  So the correct fix is going to be aborting straight after the mlookup rather than going down through the rest of the code.

@elliefm - building a Cassandane test for this will require a murder setup with source mailbox being on a backend and destination mailbox not existing.  That should reliably crash even on master based on my reading of the code.

**Comment by elliefm** [comment]
> a Cassandane test for this will require a murder setup with source mailbox being on a backend and destination mailbox not existing

Any thoughts on how to reliably set this scenario up?  If this were a replication test, I could create both folders on A and only the one folder on B.  Or I could create both folders on A, replicate to B, then delete the destination folder from B.  But with murder, I imagine any creates/deletes on one will just propagate to the other...

**Comment by elliefm** [comment]
Looks like I might be able to use `create` to create the two folders on both instances, and then `localdelete` for the destination folder on the backend instance.

**Comment by elliefm** [comment]
Test case here: https://github.com/cyrusimap/cassandane/compare/master...elliefm:issue34

... but it passes for both 2.5 and master, no sign of a crash? So I wonder what detail I've missed in the setup.

Here's the IMAP conversations during the test:

frontend:
```
[snip authentication]
C: 3 select "inbox"
S: * 0 EXISTS
S: * 0 RECENT
S: * FLAGS (\Answered \Flagged \Draft \Deleted \Seen)
S: * OK [PERMANENTFLAGS (\Answered \Flagged \Draft \Deleted \Seen \*)] Ok
S: * OK [UIDVALIDITY 1476227841] Ok
S: * OK [UIDNEXT 1] Ok
S: * OK [HIGHESTMODSEQ 3] Ok
S: * OK [URLMECH INTERNAL] Ok
S: * OK [ANNOTATIONS 65536] Ok
S: 3 OK [READ-WRITE] Completed
[snip creation of three test messages in INBOX]
C: 11 create INBOX.dest
S: 11 OK Completed
```

backend:
```
[snip authentication]
C: 3 localdelete INBOX.dest
S: 3 OK Completed
```

frontend:
```
C: 12 list "" "*"
S: * LIST (\HasChildren) "." INBOX
S: * LIST (\HasNoChildren) "." INBOX.dest
S: 12 OK Completed (0.000 secs 3 calls)
```

backend:
```
C: 4 list "" "*"
S: * LIST (\HasNoChildren) "." INBOX
S: 4 OK Completed (0.000 secs 2 calls)
```

frontend:
```
C: 13 move 1 INBOX.dest
S: 13 NO [TRYCREATE] Mailbox does not exist
C: 14 copy 2 INBOX.dest
S: 14 NO [TRYCREATE] Mailbox does not exist
```

**Comment by MASHtm** [comment]
roundcube uses
C; A0007 UID MOVE 552366 INBOX.Trash
my patched version already survives and answers
S: A0007 OK [COPYUID 1123762451 552366 766908] Completed

And INBOX.Trash is officially deleted on both front- and backend. It is the proxy code which falls through to the part I patched. It is only set as "Trash" in roundcube which fails to check the availability of it when pushing the "Trash"-Button (fixed in 1.2.2).

**Comment by elliefm** [comment]
Oh, so it's not that the folder's existence is inconsistent between frontend and backend, it's that the folder doesn't exist anywhere but the client tries to move stuff to it?  That's a whole different test then :)

**Comment by elliefm** [comment]
Crash reproduced on 2.5

**Comment by elliefm** [comment]
Adding the `!r &&` check looks like the right thing to do here -- or at least, the right thing to do short of converting the entirety of cmd_copy to use the `if (r) goto ...` idiom.

Select leaves us with either imapd_index set (if the selected mailbox is local), or backend_current set (if the selected mailbox is remote).  So the `if (!imapd_index && !backend_current) goto nomailbox;` is legit.

> So the correct fix is going to be aborting straight after the mlookup rather than going down through the rest of the code.

This is effectively what adding the `!r &&` does.  `r` was set to IMAP_MAILBOX_NONEXISTENT by the mlookup, we just need to successfully do nothing further as long as that's the case.

We segfault because we're trying to do something further: check expunge permission on (local) source mailbox. But we shouldn't be doing that, because `r` is set -- we've already failed.  We crash because we try to deref imapd_index, which in this case isn't set because the source mailbox is remote.  But even if it was set, and we didn't crash, trying to check source expunge permission at this point is bogus because we already know the destination doesn't exist and we're not doing anything.

**Comment by elliefm** [comment]
Tests and fix are now upstream (both 2.5 and master)
